### DISCUSSION TITLE
Want To Contribute In Crawling Part of FRDMS!

### DISCUSSION THREAD

**Discussion by BidyutBikashBharali** [ivan0313/Facial-Recognition-Database-Management-System#53] [General]
Hey, I'm interested to contribute in the Crawling part of your project FRDMS. So,Would you like to discuss about it ?

**Comment by ivan0313** [comment]
Hey sure! 

For the crawler from a user standpoint would be:
- can select and run preset crawlers, with argument to refine the scope of crawling
- the crawler should then collect a list of data points, each data point would have at least one picture of a person, name of that person, and other personal information
- finally all pictures collected should be run through the recognition model, automatically create a new profile if not exist
- (optional) user can create new customized crawler

My initial thoughts are the crawler would run as a task by the task worker on a thread, so user can continue to use the app. (Task worker already created for batch face recognition)

The preset crawlers would be on sites that has a collection of personals profiles and biographies, such as IMDb actors, any social networking platforms. The arguments user can input into the crawler would be for example a search keyword on Instagram, sex, age, location, occupation filters, etc, so depending on what search refinements the website provides.

Since the crawling would run as a task, the task page and task creation page would also need to be modified to accommodate for managing crawling tasks.

The last part for user to create their own crawler would be tricky tho, because the idea of the app is for user to not have to install python. But this part can be left for later.

**Reply by BidyutBikashBharali** [reply]
Hi, would you like to help me in understanding how this will work by answering my following queries:

1. Could you please clearly describe the scenario when the user will use the crawler and for what?

2. and please tell me more specifically about the arguments ( preset arguments ) which would be provided by an user to start crawling on social media platforms? What will be the main argument by which crawler will start searching on internet? I think age, sex, region etc will helps to search more specifically but not the main argument.

3. And what will be the role of image recognition model on the crawled pictures? Will it be used for grouping the matching faces and to make a profile of that face?

**Reply by ivan0313** [reply]
#### 1. 
To populate the database with profiles and images, users currently have to manually prepare a number of images and upload to FRDMS via the `/batch-rec-tasks/create` or `/facial-recognition` page, and edit the created profile manually. The crawler attempts to automate this process: populate the database with images and profiles by crawling specific websites.

#### 2. 
A generalized crawler that can be used across multiple domains of course would be ideal, but the implementation I imagine is very hard and not really what this project aims to achieve. I guess what you mean by "main arguments" would be the arguments supplied to this generalized crawler. (Of course, please do tell if know of any libraries or method to make this is achievable) 

Instead a more feasible solution is to create a number of preset crawlers that users can choose from. 

For example, we can create 4 preset crawlers for users to choose from.

```
Step 1
Choose a crawler: 
    - Instagram (https://www.instagram.com/)
    - LinkedIn (https://www.linkedin.com/)
    - IMDb (https://www.imdb.com/search/name/)
    - Biography (www.biography.com)
```

If user choose the IMDb crawler, the arguments would be 

- name
- start_birthday
- end_birthday
- birthday
- name_groups
... and more...

Since these are the search parameters provided by the website to be crawled.

![image](https://user-images.githubusercontent.com/60818397/138586495-b72c2d1a-8a71-4469-8bd4-18cb722e2480.png)

#### 3
The role image recognition model on the crawled pictures would be analysis and matching.

Analysis - The first step of FRDMS will perform on any crawled picture is to detect face in the picture, and generate the encoding for recognition. We will have to discard any data points where more than one face is detected in the crawled image, as we have no way to tell which face belongs to the textual data crawled.

Matching - yes just like what you mentioned, search the faces in the crawled picture against the existing database. If the profile of that person already exist, update the profile with the newly crawled textual & image data; if not exist, create a new profile and check if any of the unmatched faces (already in the database) belongs to this newly created profile.
